pytential/symbolic/pde/scalar.py

```diff
@@ -252,108 +342,147 @@ def map_potentials(x):  # pylint:disable=function-redefined
 
         return (
                 map_potentials(
-                    sym.S(self.kernel, inv_sqrt_w_u, **kwargs))
-                - self.alpha
-                * map_potentials(
-                    sym.D(self.kernel,
-                        sym.S(self.laplace_kernel, inv_sqrt_w_u),
-                        **kwargs)))
-
-    def operator(self, u):
-        from sumpy.kernel import HelmholtzKernel, LaplaceKernel
-
+                    sym.S(self.kernel, inv_sqrt_w_u, **kwargs)
+                    )
+                - self.alpha * map_potentials(
+                    sym.D(self.kernel, laplace_s_inv_sqrt_w_u, **kwargs)
+                    )
+                )
+
+    def operator(self, u, **kwargs):
+        """
+        :param u: symbolic variable for the density.
+        :param kwargs: additional keyword arguments passed on to the layer
+            potential constructor.
+        """
         sqrt_w = self.get_sqrt_weight()
-        inv_sqrt_w_u = cse(u/sqrt_w)
-
-        knl = self.kernel
-        lknl = self.laplace_kernel
+        inv_sqrt_w_u = sym.cse(u/sqrt_w)
+        laplace_s_inv_sqrt_w_u = sym.cse(
+                sym.S(self.laplace_kernel, inv_sqrt_w_u, qbx_forced_limit=+1)
+                )
 
-        knl_kwargs = {}
-        knl_kwargs["kernel_arguments"] = self.kernel_arguments
+        kwargs["kernel_arguments"] = self.kernel_arguments
 
-        DpS0u = sym.Dp(knl,  # noqa
-                cse(sym.S(lknl, inv_sqrt_w_u)),
-                **knl_kwargs)
+        # NOTE: the improved operator here is based on right-precondioning
+        # by a single layer and then using some Calderon identities to simplify
+        # the result. The integral equation we start with for Neumann is
+        #       I + S' + alpha D' = g
+        # where D' is hypersingular
 
         if self.use_improved_operator:
-            Dp0S0u = -0.25*u + sym.Sp(  # noqa
-                    lknl,  # noqa
-                    sym.Sp(lknl, inv_sqrt_w_u, qbx_forced_limit="avg"),
-                    qbx_forced_limit="avg")
+            def Sp(density):
+                return sym.Sp(self.laplace_kernel,
+                        density,
+                        qbx_forced_limit="avg")
 
+            # NOTE: using the Calderon identity
+            #   D' S = -u/4 + S'^2
+            Dp0S0u = -0.25 * u + Sp(Sp(inv_sqrt_w_u))
+
+            from sumpy.kernel import HelmholtzKernel, LaplaceKernel
             if isinstance(self.kernel, HelmholtzKernel):
-                DpS0u = (  # noqa
-                        sym.Dp(knl - lknl,  # noqa
-                            cse(sym.S(lknl, inv_sqrt_w_u, qbx_forced_limit=+1)),
-                            qbx_forced_limit=+1, **knl_kwargs)
+                DpS0u = (
```

Comment:
Shouldn't the same trick work for Yukawa? i.e. adding a `Dp(self.kernel - self.laplace_kernel)` to subtract out the added Laplace stuff. 

Although if `alpha = 0` is recommended for Yukawa, this makes no difference.

Reply:
It would, but there's no point to the combined-field representation for Yukawa. It'll work, but there's really no need.